conjure-java-jackson-serialization/src/main/java/com/palantir/conjure/java/serialization/CollectionValidationModule.java

```diff
@@ -0,0 +1,128 @@
+ *     <li>TODO: Values are not mutated (after deserialization)</li>
+ * </ul>
+ */
+final class CollectionValidationModule extends SimpleModule {
+
+    private static final Logger log = LoggerFactory.getLogger(CollectionValidationModule.class);
+
+    CollectionValidationModule() {
+        super("collection-validation");
+        addAbstractTypeMapping(List.class, ValidatingArrayList.class);
+        addAbstractTypeMapping(Set.class, ValidatingHashSet.class);
+        addAbstractTypeMapping(Map.class, ValidatingHashMap.class);
+    }
+
+    /** Logs noisily if null values are received. */
+    static final class ValidatingArrayList<T> extends ArrayList<T> {
+
+        @Override
+        public boolean add(T value) {
+            return super.add(logIfNull(value));
+        }
+
+        @Override
+        public void add(int index, T value) {
+            super.add(index, logIfNull(value));
+        }
+
+        @Override
+        public boolean addAll(Collection<? extends T> collection) {
+            collection.forEach(CollectionValidationModule::logIfNull);
+            return super.addAll(collection);
+        }
+
+        @Override
+        public boolean addAll(int index, Collection<? extends T> collection) {
+            collection.forEach(CollectionValidationModule::logIfNull);
+            return super.addAll(index, collection);
+        }
+
+        @Override
+        public T set(int index, T value) {
+            return super.set(index, logIfNull(value));
+        }
+    }
+
+    /** Logs noisily if null values are received. */
+    static final class ValidatingHashSet<T> extends HashSet<T> {
+
+        @Override
+        public boolean add(T value) {
+            return super.add(logIfNull(value));
+        }
+
+        @Override
+        public boolean addAll(Collection<? extends T> collection) {
+            collection.forEach(CollectionValidationModule::logIfNull);
+            return super.addAll(collection);
+        }
+    }
+
+    /** Logs noisily if null values are received. */
+    static final class ValidatingHashMap<K, V> extends HashMap<K, V> {
+        private static final BiConsumer<Object, Object> validator = (key, value) -> logIfNull(value);
+
+        @Override
+        public V put(K key, V value) {
+            return super.put(key, logIfNull(value));
+        }
+
+        @Override
+        public void putAll(Map<? extends K, ? extends V> map) {
+            map.forEach(validator);
+            super.putAll(map);
+        }
+    }
+
+    @Nullable
+    private static <T> T logIfNull(@Nullable T value) {
+        // Log noisily in production
+        if (value == null) {
+            log.warn("Detected a null value, null values are not allowed by Conjure", new IllegalArgumentException());
+        }
+        // Tests should fail, but we're not ready to cause breaks in deployed systems yet.
+        assert value != null : "Null values are not allowed by Conjure";
```

Comment:
This may be contentious, interested in feedback.

I'd like tests which rely on unexpected behavior to fail rather than making their way to production and filing the logs -- the added logging should be considered a fallback.

Reply:
This is a pretty cool solution :) 

I think we'd need to excavate something out (or bake something into a plugin) to make sure tests run with `-ea`, because I just tried putting `assert false` in a test and it still passed!

Reply:
I think our intellij configuration is wonky, if gradle _ever_ runs tests without assertions enabled, that's a huge bug.